contracts/Smt.metasol

```diff
@@ -156,6 +165,69 @@ library SMT{{hashname}} {
     function inc(uint8 x) pure private returns (uint8) {
         unchecked { return x + 1; }
     }
+
+    function modifyElement(MerkleTree storage tree, uint index, bytes memory data) internal {
+        require(tree.elementData[index].length != 0, "Element doesn't exists");
+        _modifyElement(tree, index, data);
+    }
+
+    {{ if SET }}
+    function addToNextEmpty(MerkleTree storage tree, bytes memory data) internal returns (uint256) {
+        uint256 index = tree.firstEmptySlot;
+        tree.firstEmptySlot += 1;
+
+        require(index < (1 << tree.maxDepth), "Index out of range");
+        if (index == (1 << tree.depth)) {
+            _increaseDepth(tree, 1);
+        }
+
+        _addElement(tree, index, data);
+        return index;
+    }
+
+    function removeAndRebase(MerkleTree storage tree, uint index) internal returns (bool){
+        uint256 localFirstEmptySlot = tree.firstEmptySlot - 1;
+        tree.firstEmptySlot = localFirstEmptySlot;
```

Comment:
why do we need the auxiliary parameter?

Reply:
to save gas. If you read local variable it's cost less gas than when you read storage

Reply:
IMHO, the gas usage is increased in this case than:
`tree.firstEmptySlot = tree.firstEmptySlot - 1`

The number of op codes in your case are more than number of op codes than the code above. Certainly, the difference is small, but I think in terms of gas usage it is not optimum.
The image below shows a simple test result ( compiled without optimization enabled):
uint public a = 3;
    
    function gasTest1() public{
        uint temp = a - 1;
        a = temp;
    }
    
    function gasTest2() public{
        a = a - 1;
    }
![image](https://user-images.githubusercontent.com/67259426/144632784-20db369e-b860-4bac-9591-5a291a8b3c9d.png)

Reply:
but after that i read from localFirstEmptySlot few times

```
    function removeAndRebase(MerkleTree storage tree, uint index) internal returns (bool){
        uint256 localFirstEmptySlot = tree.firstEmptySlot - 1;
        tree.firstEmptySlot = localFirstEmptySlot;
        bool result = false;
        if (index != localFirstEmptySlot) {
            // move the last element into empty place
            bytes memory element = tree.elementData[localFirstEmptySlot];
            tree.modifyElement(index, element);
            result = true;
        }
        _removeElement(tree, localFirstEmptySlot);

        if ((localFirstEmptySlot == (1 << (tree.depth - 1))) && (localFirstEmptySlot > 1)) {
            _decreaseDepth(tree, 1);
        }

        return result;
    }
```

Reply:
You are right. I did not follow the rest of code.